src/occupation.jl

```diff
@@ -64,50 +118,24 @@ function compute_fermi_level(basis::PlaneWaveBasis{T}, eigenvalues;
+    @assert objective(min_ε) < 0 < objective(max_ε)
+    Roots.find_zero(objective, (min_ε, max_ε), Roots.Bisection(), atol=eps(T))
+end
 
-        # For zero temperature, two cases arise: either there are as many bands
-        # as electrons, in which case we set εF to the highest energy level
-        # reached, or there are unoccupied conduction bands and we take
-        # εF as the midpoint between valence and conduction bands.
-        if compute_n_elec(max_ε) ≈ n_electrons
-            εF = max_ε
-        else
-            # The sanity check above ensures that n_fill is well defined
-            n_fill = div(n_electrons, n_spin * filled_occ, RoundUp)
-            # highest occupied energy level
-            HOMO = maximum([εk[n_fill] for εk in eigenvalues])
-            HOMO = mpi_max(HOMO, basis.comm_kpts)
-            # lowest unoccupied energy level, be careful that not all k-points
-            # might have at least n_fill+1 energy levels so we have to take care
-            # of that by specifying init to minimum
-            LUMO = minimum(minimum.([εk[n_fill+1:end] for εk in eigenvalues]; init=T(Inf)))
-            LUMO = mpi_min(LUMO, basis.comm_kpts)
-            εF = (HOMO + LUMO) / 2
-        end
-    else
-        # For finite temperature, just use bisection; note that with MP smearing
-        # there might be multiple possible Fermi levels. This could be sped up
-        # with more advanced methods (e.g. false position), but more care has to
-        # be taken with convergence criteria and the like
-        @assert compute_n_elec(min_ε) < n_electrons < compute_n_elec(max_ε)
-        εF = Roots.find_zero(εF -> compute_n_elec(εF) - n_electrons, (min_ε, max_ε),
-                             Roots.Bisection(), atol=eps(T))
+function compute_fermi_level(basis::PlaneWaveBasis, eigenvalues ::GaussianNewton;
+                             temperature, smearing)
+    if iszero(temperature)
+        return compute_fermi_level(basis, eigenvalues, ZeroTemperature())
+    elseif smearing in (Smearing.Gaussian(), Smearing.FermiDirac())
+        # Monotonous smearing functions ... only one Fermi level anyway
+        return compute_fermi_level(basis, eigenvalues, Bisection(); temperature, smearing)
     end
 
-    if !isapprox(compute_n_elec(εF), n_electrons)
-        if iszero(temperature)
-            error("Unable to find non-fractional occupations that have the " *
-                  "correct number of electrons. You should add a temperature.")
-        else
-            error("This should not happen, debug me!")
-        end
-    end
-    εF
+    εF_guess = compute_fermi_level(basis, eigenvalues, Bisection();
+                                   temperature, smearing=Smearing.Gaussian())
+
+    n_elec(εF)  = deviation_n_electrons(basis, eigenvalues, εF; smearing, temperature)
+    dn_elec(εF) = ForwardDiff.derivative(objective, εF)
+    Roots.find_zero((n_elec, dn_elec), εF_guess, Newton())
```

Comment:
See the QE implementation [here](https://github.com/QEF/q-e/blob/9ddde6fa44e0fb1c7debb78532a4b581a6ef82fd/PW/src/efermig.f90#L118).

To me, it seems like they do the Newton's minimization on
```julia
f(eF) = (compute_n_electrons(basis, eigenvalues, eF; smearing, temperature) - n_electrons)^2
df(eF) = ForwardDiff.derivative(f, eF)
ddf(eF) = ForwardDiff.derivative(df, eF)

Roots.find_zero((df, ddf), eF_Gaussian, Newton())
```

This is what I implemented for my project, and it seems to perform better and fail less than Newton's minimization on `(n, dn)`. I haven't given any thought to why this is the case or why Flaviano chose to do it this way, but it would be good to ask.

Reply:
At least certainly even on simple things the quadratic one performs a lot better.

I guess because the route to the minimum is steeper.

Reply:
What a weird thing to do. Suppose you want to solve for f(x)=0, then the standard newton method requires f' to be nonzero. If instead you set g = f^2 and solve for g' = 2 f' f = 0, that requires g'' = 2f'' f + 2f'^2 = 0. Since f is zero at the solution, that's f'^2. So if f' is small, then f'^2 is even smaller, so that's just worse? So they don't actually solve for f = 0, but rather try to minimize f^2. In your plot in the other thread it looked as if f was small but not exactly zero at the "right" fermi level, correct?

Reply:
Yeah, but I interpreted that as a result of the tolerance they use for MV and MP, which is an absolute error in the number of electrons of 1e-2 (which seems really big to me). For comparison, they use 1e-10 as the tolerance for the bisection method.

Reply:
In my experiments I found that the actual numerical consistency for the cases with multiple Fermi solutions can be pretty bad in any case. I.e. you find a solution and then still DFTK's warning on the electron count for the density flags up.

But I agree a trustworthy approach should find a root for f = 0 and should ensure the electron count is as it should be, certainly to a higher accuracy than 1e-2.

> In your plot in the other thread it looked as if f was small but not exactly zero at the "right" fermi level, correct?

Not sure what you mean here @antoine-levitt. f is zero at the fermi level candidates, but f' can be all over the place ... generally the "best" seems to be the one where it is smallest from my current understanding, though.

Reply:
> In my experiments I found that the actual numerical consistency for the cases with multiple Fermi solutions can be pretty bad in any case. I.e. you find a solution and then still DFTK's warning on the electron count for the density flags up.

This actually was a bug.